0-module-and-module-source.emu

```diff
@@ -537,27 +569,48 @@ location: https://tc39.es/proposal-compartments/
                 1. If Type(_moduleSource_) is not Object, throw a *TypeError* exception.
-                1. If IsCallable(_importHook_) is *false*, throw a *TypeError* exception.
-                1. If _importMeta_ is *undefined*, then
-                  1. Set _importMeta_ to *null*.
-                1. Else, if Type(_importMeta_) is not Object, then
-                  1. Throw a *TypeError* exception.
+                _importHook_
                 1. Perform ? RequireInternalSlot(_moduleSource_, [[ModuleSource]]).
-                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%Module.prototype%"*, « [[Module]], [[ModuleSourceInstance]], [[ImportHook]] »).
                 1. Let _moduleRecord_ to ! CreateModuleRecord(_moduleSource_.[[ModuleSource]]).
-                1. Set _moduleRecord_.[[ModuleInstance]] to _O_.
-                1. Set _moduleRecord_.[[ImportMeta]] to _importMeta_.
-                1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _importHook_ and _referral_ and performs the following steps when called
-                  1. Assert: Type(_specifier_) is String.
-                  1. return ? Call(_importHook_, *undefined*, « _specifier_, _referral_ »).
-                1. Set _O_.[[Module]] to _moduleRecord_.
+                1. If Type(_handler_) is Object, then
+                  1. If _handler_ is *null, throw a TypeError exception.
+                  1. Let _importHook_ be ? GetMethod(handler, "importHook").
+                  1. If _importHook_ is *undefined*, then
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _moduleRecord_ and performs the following steps when called
+                      1. Assert: Type(_specifier_) is String.
+                      1. NOTE: Default import hook.
+                      1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
+                      1. Perform HostImportModuleDynamically(_moduleRecord_, _specifier_, _promiseCapability_).
+                      1. Return _promiseCapability_.[[Promise]].
+                  1. Else,
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _importHook_ and _handler_ and performs the following steps when called
+                      1. Assert: Type(_specifier_) is String.
+                      1. return ? Call(_importHook_, _handler_, « _specifier_ »).
+                  1. Let _importMetaHook_ be ? GetMethod(handler, "importMetaHook").
+                  1. If _importMetaHook_ is *undefined*, then
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_importMeta_) that captures _moduleRecord_ and performs the following steps when called
+                      1. NOTE: Default import.meta hook.
+                      1. Let _importMetaValues_ be HostGetImportMetaProperties(_moduleRecord_).
```

Comment:
This whole block is the same as in the import.meta runtime semantics, maybe we can abstract it out.

Reply:
@nicolo-ribaudo I did just that... and it is more simple now... eventually, after the refactor... can we do the same for the importHook? That would simplify this even more.

Reply:
Actually, I went ahead and made the change for import hook as well to be undefined. It is looking good IMO. Thanks for the suggestion.